src/audio/mux/mux.c

```diff
@@ -414,18 +416,31 @@ static int mux_copy(struct comp_dev *dev)
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
+	buffer_lock(sink, &flags);
+
 	/* check if sink is active */
-	if (sink->sink->state != dev->state)
+	if (sink->sink->state != dev->state) {
+		for (i = 0; i < MUX_MAX_STREAMS; i++) {
+			if (!sources[i])
+				continue;
+			buffer_unlock(sources[i], flags);
+		}
```

Comment:
where is the corresponding buffer_lock for this? Code looks strange.

Reply:
@dbaluta I think it's the one on line 403 (it's in the same function).

Reply:
I see. thanks.

Reply:
Yeah, I'm not seeing this, I can see lock(source) and unlock(sources[i]) which seem unaligned as they are different objects. Some inline comments may help here.

Reply:
@lgirdwood They are the same objects.

Reply:
@tlauda this looks error prone.

Reply:
@dbaluta What do you mean?

Reply:
I wonder if simplification would be better here. i.e.
```c
	/* align source streams with their respective configurations */
	list_for_item(clist, &dev->bsource_list) {
		source = container_of(clist, struct comp_buffer, sink_list);
		buffer_lock(source, &flags);
		if (source->source->state == dev->state) {
			num_sources++;
			i = get_stream_index(cd, source->pipeline_id);
			sources[i] = source;
			sources_stream[i] = &source->stream;
		} else {
			buffer_unlock(source, flags);
		}
	}
```
Means we only have certain source buffers locked, i.e only the buffers that out configuration is telling us to copy from. Would it make more sense to just lock all sources and then unlock all sources ?

Another thing that we should do is to change the buffer locks to reader and writer locking.
```
buffer_rlock(buffer, flags);
buffer_wlock(buffer, flags);
```
This way we only lock a buffer for reading or writing (i.e.reader lock updates r_ptr and avail, writer lock updates w_ptr and free). This would allow cross core buffers that would not block a core that the current implementation might.

Reply:
@lgirdwood It's definitely candidate for future improvement. This patch is about aligning with rest of the components.

Reply:
Ok, great for the r/w buffer locking update.